internal/fs/handle/file.go

```diff
@@ -221,3 +272,57 @@ func (fh *FileHandle) tryEnsureReader(ctx context.Context, sequentialReadSizeMb
 	fh.reader = rr
 	return
 }
+
+// If possible, ensure that fh.readManager is set to an appropriate read manager
+// for the current state of the inode otherwise set it to nil.
+//
+// LOCKS_REQUIRED(fh)
+// LOCKS_REQUIRED(fh.inode)
+func (fh *FileHandle) tryEnsureReadManager(ctx context.Context, sequentialReadSizeMb int32) error {
+	// Ensure inode content is up-to-date. If content caching is enabled,
+	// this might make the inode's source generation non-authoritative.
+	if err := fh.inode.CacheEnsureContent(ctx); err != nil {
+		return fmt.Errorf("failed to ensure inode content: %w", err)
+	}
+
+	// If the inode is dirty (its source generation isn't authoritative),
+	// we cannot create a read manager based on the cloud object.
+	// Invalidate any existing read manager and return.
+	if !fh.inode.SourceGenerationIsAuthoritative() {
+		fh.destroyReadManager()
+		return nil
+	}
+
+	// If a read manager already exists and its object generation matches the inode's
+	// source generation, it's still valid. Just update its size and return.
```

Comment:
I think, this size update is required only for zonal bucket, good to convey in the comment.